lib/elixir/lib/task.ex

```diff
@@ -768,35 +786,44 @@ defmodule Task do
         end
 
         def init(_opts) do
-          # We will keep all running tasks in a map
-          {:ok, %{tasks: %{}}}
+          # We will keep the refs and tags for each task in maps.
+          {:ok, %{task_refs: %{}, task_tags: %{}}}
         end
 
         # Imagine we invoke a task from the GenServer to access a URL...
         def handle_call(:some_message, _from, state) do
-          url = ...
+          url = "..."
           task = Task.Supervisor.async_nolink(MyApp.TaskSupervisor, fn -> fetch_url(url) end)
 
-          # After we start the task, we store its reference and the url it is fetching
-          state = put_in(state.tasks[task.ref], url)
+          # After we start the task, we store its reference and the URL it is fetching
+          state = put_in(state.task_refs[task.ref], url)
+          # We also store its tag and map it to the reference
+          state = put_in(state.task_tags[task.tag], task.ref)
 
           {:reply, :ok, state}
         end
 
         # If the task succeeds...
-        def handle_info({ref, result}, state) do
-          # The task succeed so we can demonitor its reference
+        def handle_info({tag, result}, state) do
+          ref = Map.fetch!(state.task_tags, tag)
+
+          # The task succeeded so we can demonitor its reference.
           Process.demonitor(ref, [:flush])
```

Comment:
I think we don't need this, given `Task` itself demonitors (with `alias: reply_demonitor`)?

Reply:
Good call.